cryoet_data_portal_neuroglancer/models/shader_builder.py

```diff
@@ -0,0 +1,149 @@
+"""Create GLSL shaders for Neuroglancer."""
+
+from typing import Any, Optional
+
+from cryoet_data_portal_neuroglancer.utils import get_window_limits_from_contrast_limits
+
+TAB = "  "
+
+
+class ShaderBuilder:
+    def __init__(self):
+        self._shader_pre_main = ""
+        self._shader_main_function = ""
+        self._shader_controls = {}
+
+    def add_to_shader_controls(self, code: str | list[str]):
+        if isinstance(code, str):
+            self._shader_pre_main += code
+        else:
+            self._shader_pre_main += "\n".join(code)
+        self._shader_pre_main += "\n"
+        return self
+
+    def add_to_shader_main(self, code: str | list[str], indent: int = 1):
+        if isinstance(code, str):
+            self._shader_main_function += TAB * indent + code
+        else:
+            self._shader_main_function += "\n".join([TAB * indent + line for line in code])
+        self._shader_main_function += "\n"
+        return self
+
+    def _make_main(self) -> str:
+        return f"void main() {{\n{self._shader_main_function}}}"
+
+    def _make_pre_main(self) -> str:
+        """Sort the preamble for more visually appealing code"""
+        # Extract all the #uicontrol lines and put them at the top
+        uicontrol_lines = []
+        pre_main_lines = []
+        for line in self._shader_pre_main.split("\n"):
+            if line.startswith("#uicontrol"):
+                uicontrol_lines.append(line)
+            else:
+                pre_main_lines.append(line)
+        # Create a blank line between the uicontrols and the rest of the code
+        if len(pre_main_lines) > 1:
+            pre_main_lines.insert(0, "")
+        return "\n".join(uicontrol_lines + pre_main_lines)
+
+    def build_shader(self) -> dict[str, str | dict[str, Any]]:
+        return {
+            "shader": self._make_pre_main() + "\n" + self._make_main(),
+            "shaderControls": self._shader_controls,
+        }
+
+    def make_invlerp_component(
+        self,
+        name: str,
+        contrast_limits: tuple[float, float],
+        window_limits: tuple[float, float],
+    ) -> str:
+        controls = self._shader_controls.setdefault(name, {})
+        controls["range"] = list(contrast_limits)
+        controls["window"] = list(window_limits)
+        return f"#uicontrol invlerp {name}"
+
+    def make_invertible_invlerp_component(
+        self,
+        name: str,
+        contrast_limits: tuple[float, float],
+        window_limits: tuple[float, float],
+    ) -> list[str]:
+        invlerp_component = self.make_invlerp_component(name, contrast_limits, window_limits)
+        checkbox_part = f"#uicontrol bool invert_{name} checkbox"
+        data_value_getter = [
+            f"float get_{name}()" + " {",
+            f"{TAB}return invert_{name} ? 1.0 - {name}() : {name}();",
```

Comment:
As the values can be outside the range of 0 and 1, should the inversion of values also reflect that?

Reply:
Thanks for spotting this one, but I think it should be ok! This is because:

In the first step, the invlerp GLSL shader function will take the raw data value of the underlying image, and map it to a value based on the contrast limits. So for certain, the value after this processing might lie outside the range 0 - 1, unless we specify to clamp the data value in the invlerp on line 65. E.g. if the image value from the zarr was 100, and the contrast limits were 0 to 50, we'd invlerp to 2.0.

However, the regular invlerp function is `normal = (inputValue - lowerBound) / (upperBound - lowerBound)`. The inverted invlerp function, which we'd want to calculate when the checkbox is ticked, is `inverted = (inputValue - upperBound) / (lowerBound - upperBound)`. As it happens, this means that `1.0 - normal = inverted`, and we are ok, even if the values are outside the range 0 - 1.

At this point, values outside the range 0 - 1 might get passed to neuroglancer functions which usually work with values in the range 0 - 1, but neuroglancer does clamping under the hood to deal with these cases (and it's quite common that these get passed to neuroglancer)

Reply:
Hope this answers your question @manasaV3 , but please let me know if I have misunderstood or missed something!